Vungle/VungleRouter.m

```diff
@@ -588,44 +624,78 @@ - (VungleAdSize)getVungleBannerAdSizeType:(CGSize)size
+                                             eventID:(NSString *)eventID
+                                     withBannerState:(BannerRouterDelegateState)state
+{
+    id<VungleRouterDelegate> targetDelegate = [self getFullScreenDelegateWithPlacement:placementID
+                                                                               eventID:eventID];
+    if (!targetDelegate) {
+        targetDelegate = [self getBannerDelegateWithPlacement:placementID eventID:eventID withBannerState:state];
+    }
+    
+    return targetDelegate;
+}
+
+- (id<VungleRouterDelegate>)getFullScreenDelegateWithPlacement:(NSString *)placementID
+                                                       eventID:(NSString *)eventID
+{
     if (!placementID.length) {
         return nil;
     }
+    
+    id<VungleRouterDelegate> targetDelegate;
+    if (eventID.length > 0) {
+        targetDelegate = [self.hbDelegatesDict objectForKey:eventID];
+    } else {
+        targetDelegate = [self.delegatesDict objectForKey:placementID];
+    }
+    
+    return targetDelegate;
+}
 
-    @synchronized (self) {
-        id<VungleRouterDelegate> targetDelegate = [self.delegatesDict objectForKey:placementID];
-        if (!targetDelegate) {
-            id<VungleRouterDelegate> bannerDelegate =
-            [self getBannerDelegateWithPlacement:placementID
-                                 withBannerState:state];
-            if (bannerDelegate) {
-                targetDelegate = bannerDelegate;
-            }
-        }
-        return targetDelegate;
+- (id<VungleRouterDelegate>)getBannerDelegateWithPlacement:(NSString *)placementID
+                                                       eventID:(NSString *)eventID
+{
+    if (!placementID.length) {
+        return nil;
     }
+    
+    id<VungleRouterDelegate> targetDelegate;
+    if (eventID.length > 0) {
+        targetDelegate = [self.hbBannerDelegates objectForKey:eventID];
+    } else {
+        targetDelegate = [self.bannerDelegates objectForKey:placementID];
+    }
+    
+    return targetDelegate;
 }
 
 - (id<VungleRouterDelegate>)getBannerDelegateWithPlacement:(NSString *)placementID
+                                                   eventID:(NSString *)eventID
                                            withBannerState:(BannerRouterDelegateState)state
 {
-    id<VungleRouterDelegate> bannerDelegate =
-                [self.bannerDelegates objectForKey:placementID];
-    if (bannerDelegate && bannerDelegate.bannerState == state) {
-        return bannerDelegate;
+    id<VungleRouterDelegate> targetDelegate = [self getBannerDelegateWithPlacement:placementID eventID:eventID];
+    if (targetDelegate.bannerState != state) {
+        return nil;
     }
-    return nil;
+
+    return targetDelegate;
 }
 
 - (void)replaceOldBannerDelegateWithDelegate:(id<VungleRouterDelegate>)delegate
-                             withPlacementID:(NSString *)placementID
```

Comment:
Just thought of this:
I wonder if this method is still applicable for header bidding.
Every bid has their own unique id so how would it be possible to know and replace the old banner delegate using the event id? Using the placement id cannot be done because the publisher would be able to load and play multiple ads of the same banner placement (right?). How would the delegate understand which is the older one?

Reply:
@stanley-vungle Yes, I also think this method is applicable for header bidding. We need to use eventId to differentiate multiple ads of the same banner placement. 😊